subsys/usb/usb_device.c

```diff
@@ -503,18 +513,53 @@ static bool set_endpoint(const struct usb_ep_descriptor *ep_desc)
 		ep_cfg.ep_addr, ep_cfg.ep_type, ep_cfg.ep_mps);
 
 	if (usb_dc_ep_configure(&ep_cfg) < 0) {
-		LOG_WRN("Failed to configure endpoint %x", ep_cfg.ep_addr);
+		return false;
 	}
 
 	if (usb_dc_ep_enable(ep_cfg.ep_addr) < 0) {
-		LOG_WRN("Failed to enable endpoint %x", ep_cfg.ep_addr);
+		return false;
 	}
 
 	usb_dev.configured = true;
 
 	return true;
 }
 
+/*
+ * @brief Reset configuration and disable endpoint
+ *
+ * This function resets endpoint configuration to
+ * default values and disables it.
+ *
+ * @param [in]  ep_desc Endpoint descriptor byte array
+ *
+ * @return true if successfully deconfigured and disabled
+ */
+static bool reset_endpoint(const struct usb_ep_descriptor *ep_desc)
+{
+	struct usb_dc_ep_cfg_data ep_cfg = {0};
+
+	ep_cfg.ep_addr = ep_desc->bEndpointAddress;
+	ep_cfg.ep_type = ep_desc->bmAttributes;
+
+	if (ep_desc->bmAttributes > USB_DC_EP_INTERRUPT) {
+		return false;
+	}
+
+	LOG_DBG("Reset endpoint 0x%x type %u",
+		ep_cfg.ep_addr, ep_cfg.ep_type);
+
+	if (usb_dc_ep_configure(&ep_cfg) < 0) {
```

Comment:
I think just `usb_dc_ep_disable` should be enough, also calling usb_dc_ep_configure with ep_mps = 0 would not work/break something in the drivers and execute a lot of unnecessary code. I guess even usb_dc_ep_disable would not work out of the box and we have to test and fix it for every driver. I will help you here.